torch/nn/utils/prune.py

```diff
@@ -89,119 +91,116 @@ def apply(cls, module, name, *args, **kwargs):
-            # combine `methods` with `old_method`, if `old_method` exists
-            if old_method is not None:  # meaning that there was a hook
-                # if the hook is already a pruning container, just add the
-                # new pruning method to the container
-                if isinstance(old_method, PruningContainer):
-                    old_method.add_pruning_method(method)
-                    method = old_method  # rename old_method --> method
-
-                # if the hook is simply a single pruning method, create a
-                # container, add the old pruning method and the new one
-                elif isinstance(old_method, BasePruningMethod):
-                    container = PruningContainer(old_method)
-                    # Have the pruning method remember the name of its tensor
-                    # setattr(container, '_tensor_name', name)
-                    container.add_pruning_method(method)
-                    method = container  # rename container --> method
-            return method
-
-        method = _get_composite_method(cls, module, name, *args, **kwargs)
-        # at this point we have no forward_pre_hooks but we could have an
-        # active reparametrization of the tensor if another pruning method
-        # had been applied (in which case `method` would be a PruningContainer
-        # and not a simple pruning method).
-
-        # Pruning is to be applied to the module's tensor named `name`,
-        # starting from the state it is found in prior to this iteration of
-        # pruning
-        orig = getattr(module, name)
-
-        # If this is the first time pruning is applied, take care of moving 
-        # the original tensor to a new parameter called name + '_orig' and
-        # and deleting the original parameter
-        if not isinstance(method, PruningContainer):
-            # copy `module[name]` to `module[name + '_orig']`
-            module.register_parameter(name + "_orig", orig)
-            # temporarily delete `module[name]`
-            del module._parameters[name]
-            default_mask = torch.ones_like(orig)  # temp
-        # If this is not the first time pruning is applied, all of the above
-        # has been done before in a previous pruning iteration, so we're good
-        # to go
-        else:
-            default_mask = getattr(module, name + "_mask").detach().clone(memory_format=torch.contiguous_format)
-
-        # Use try/except because if anything goes wrong with the mask 
-        # computation etc., you'd want to roll back.
+        # If the last two methods (this one and the previous one) were Pruning methods, we merge them
+        param = parametrize.get_forward_parametrizations(module, name)
+
+        # We do not allow to put forward Parametrizations and PruningMethods at the same time,
+        # because it collides with the merge of the previous masks into PruningContainers.
+        # We could support this, but I don't think we want to
```

Comment:
If users want to apply a parametrization that is logically compatible with pruning, they should be able to

Reply:
This comment is outdated, I will change it. We do not support several forward parametrizations at the same time. I thought that these could be supported, but by the very nature of how forward hooks at a module level work, there is no obvious way to chain them.